src/modules/Update.rb

```diff
@@ -132,6 +132,55 @@ def ListOfRegexpsMatchesProduct(regexp_items, product)
       ret
     end
 
+    # Set vendors which are compatilbe (defined in the product description).
+    # Do not report an error if there is a vendor change while
+    # the update between these defined vendors and products.
+    def SetCompatibleVendors
+      return unless Mode.update # Update mode only
+
+      upgrade_settings = ProductFeatures.GetFeature("software", "upgrade")
+      return unless upgrade_settings
+
+      product_upgrades = upgrade_settings["product_upgrades"]
+      return unless product_upgrades
+
+      unless product_upgrades.is_a?(Array)
+        log.error("software/upgrade/product_upgrade in control.xml is not an array")
+        return
+      end
+
+      product_upgrades.each do |product_upgrade|
+        log.info("Set compatible vendors defined in control.xml: #{product_upgrade}")
+
+        if !product_upgrade.is_a?(Hash) ||
+            !product_upgrade["compatible_vendors"] ||
+            !product_upgrade["compatible_vendors"].is_a?(Array)
+          log.error("compatible_vendors are not defined.")
+          next
+        end
+        if !product_upgrade["from"]
```

Comment:
The log message is a bit unclear and I think we can make the check simpler, the `control.xml` files are validated against the schema so it's very unlikely that this mandatory value will be missing. Let's use some simpler code like:

```ruby
if !product_upgrade["from"] || !product_upgrade["to"]
  log.error("The vendor setting does not define 'from' or 'to' value: #{product_upgrade.inspect}")
  next
end
```